File: othello_python/board.py
"""
Class Board
Othello (Reversi) game implementation
https://en.wikipedia.org/wiki/Reversi
Akseli Lukkarila
2019
"""
from typing import List, Optional, Tuple

from colorprint import Color, get_color, print_color
from utils import Disk, Move, Square


class Board:
    """Handles game board state and logic."""

    STEP_DIRECTIONS = (
        (0, -1),
        (0, 1),
        (1, 0),
        (-1, 0),
        (1, -1),
        (1, 1),
        (-1, 1),
        (-1, -1),
    )

    def __init__(self, size=8):
        self._size = size
        # init game board with empty disks
        self._board = [[Disk.EMPTY for _ in range(self._size)] for _ in range(self._size)]

        # set starting positions
        row = (self._size - 1) // 2 if self._size % 2 == 0 else (self._size - 1) // 2 - 1
        col = self._size // 2
        self._board[row][row] = Disk.BLACK
        self._board[row][col] = Disk.WHITE
        self._board[col][row] = Disk.WHITE
        self._board[col][col] = Disk.BLACK

        # keep track of empty squares on board to avoid checking already filled positions
        self._empty_squares = set(
            Square(x, y) for x in range(self._size) for y in range(self._size) if self._board[y][x] == Disk.EMPTY
        )

    def can_play(self) -> bool:
        """Return true if board contains empty squares -> still possible to make a move."""
        return bool(self._empty_squares)

    def place_disk(self, move: Move) -> None:
        """Update board for given disk placement."""
        start = move.square
        assert self._get_square(start) == Disk.EMPTY, f"Trying to place disk to an occupied square {move.square}!"
        self._set_square(start, move.disk)
        self._empty_squares.remove(start)
        for step in move.directions:
            pos = start + step
            while self._get_square(pos) == move.disk.other_disk():
                self._set_square(pos, move.disk)
                pos += step

    def possible_moves(self, disk: Disk) -> List[Move]:
        """Returns a list of all possible moves for the given disk color."""
        moves = []
        for square in self._empty_squares:
            value = 0
            directions = []
            # try stepping in all directions from this starting point
            for step in self.STEP_DIRECTIONS:
                pos = square + step
                # not valid if next disk in line is own
                if self._get_square(pos) == disk:
                    continue
                steps = 0
                # keep stepping over opponents disks
                while self._get_square(pos) == disk.other_disk():
                    steps += 1
                    pos += step
                # successful move if this direction ends in own disk
                if self._get_square(pos) == disk:
                    value += steps
                    directions.append(step)

            if value:
                moves.append(Move(square, disk, value, directions))

        return sorted(moves)

    def print_moves(self, moves: List[Move]):
        """Print available move coordinates and resulting points gained."""
        print_color(f"  Possible moves ({len(moves)}):", Color.yellow)
        # convert board from Disk enums to strings
        board = [[disk.board_char() for disk in row] for row in self._board]
        for move in moves:
            print(f"  {move}")
            x, y = move.square
            # add move value to matching position on board
            board[y][x] = get_color(move.value, Color.yellow)

        # print board with move positions
        print(f"    {' '.join(get_color(str(x), bold=True) for x in range(self._size))}")
        for index, row in enumerate(board):
            text = get_color(f"  {index} ", bold=True)
            text += " ".join(disk for disk in row)
            print(text)

    def print_score(self):
        """Count and print the number of black and white disks."""
        black, white = self._player_scores()
        print(f"\n{self}")
        print(f"Score: {get_color(black, Disk.BLACK.color())} | {get_color(white, Disk.WHITE.color())}")

    def result(self) -> Disk:
        """Calculates the final score and returns the winning player disk."""
        total_score = self._score()
        if total_score == 0:
            return Disk.EMPTY

        return Disk.WHITE if total_score > 0 else Disk.BLACK

    def _check_coordinates(self, x: int, y: int) -> bool:
        """Check that the given coordinates are inside the board."""
        return 0 <= x < self._size and 0 <= y < self._size

    def _player_scores(self) -> Tuple[int, int]:
        """Count and return the number of black and white disks (white, black)."""
        white = 0  # sum(1 for d in self._board if d == Disk.WHITE)
        black = 0  # sum(1 for d in self._board if d == Disk.BLACK)
        for row in self._board:
            for disk in row:
                if disk == Disk.WHITE:
                    white += 1
                elif disk == Disk.BLACK:
                    black += 1

        return black, white

    def _score(self) -> int:
        """Returns the total score (positive means more white disks and negative means more black disks)."""
        return sum(sum(row) for row in self._board)

    def _get_square(self, square: Square) -> Optional[Disk]:
        """Returns the state of the board (empty, white, black) at the given square."""
        x, y = square
        return self._board[y][x] if self._check_coordinates(x, y) else None

    def _set_square(self, square: Square, disk: Disk):
        """Sets the given square to given value."""
        x, y = square
        if not self._check_coordinates(x, y):
            raise ValueError(f"Invalid coordinates {square}!")
        self._board[y][x] = disk

    def __str__(self):
        """Format game board to string."""
        text = f"  {' '.join(get_color(str(x), bold=True) for x in range(self._size))}"
        for index, row in enumerate(self._board):
            text += get_color(f"\n{index} ", bold=True)
            text += " ".join(disk.board_char() for disk in row)
        return text

File: swift/Sources/othello_swift/board.swift
//==========================================================
// Board
// Defines the game board
// Akseli Lukkarila
// 2019-2024
//==========================================================

/// Handles game board state and logic.
class Board {
    let size: Int
    var board: [Disk]
    var emptySquares = Set<Square>()
    let stepDirections = [[0, -1], [0, 1], [1, 0], [-1, 0], [1, -1], [1, 1], [-1, 1], [-1, -1]]

    init(size: Int) {
        self.size = size
        // init game board with empty disks
        self.board = [Disk](repeating: Disk.empty, count: size * size)

        // set starting positions
        let pos = (self.size - 1) / 2
        let row: Int = self.size % 2 == 0 ? pos : pos - 1
        let col: Int = self.size / 2
        self.board[row * self.size + row] = Disk.white
        self.board[row * self.size + col] = Disk.black
        self.board[col * self.size + row] = Disk.black
        self.board[col * self.size + col] = Disk.white

        // keep track of empty squares on board to avoid checking already filled positions
        for y in 0 ..< self.size {
            for x in 0 ..< self.size {
                if self.board[y * self.size + x] == Disk.empty {
                    self.emptySquares.insert(Square(x, y))
                }
            }
        }
    }

    /// Return true if board contains empty squares.
    func canPlay() -> Bool {
        !self.emptySquares.isEmpty
    }

    /// Update board for given disk placement.
    func placeDisk(move: Move) {
        let start = move.square
        self.setSquare(start, move.disk)
        self.emptySquares.remove(start)
        for step in move.directions {
            var pos = start + step
            while self.getSquare(pos) == move.disk.opponent() {
                self.setSquare(pos, move.disk)
                pos += step
            }
        }
    }

    /// Returns a list of possible moves for given player.
    func possibleMoves(disk: Disk) -> [Move] {
        var moves = [Move]()
        let other: Disk = disk.opponent()
        for square in self.emptySquares {
            var value = 0
            var directions = [Square]()
            for dir in self.stepDirections {
                let step = Square(dir[0], dir[1])
                var pos = square + step
                // next square in this directions needs to be opponents disk
                if self.getSquare(pos) != other {
                    continue
                }
                var steps = 0
                // keep stepping forward while opponents disks are found
                while self.getSquare(pos) == other {
                    steps += 1
                    pos += step
                }
                // valid move if a line of opponents disks ends in own disk
                if self.getSquare(pos) != disk {
                    continue
                }
                value += steps
                directions.append(step)
            }
            if value > 0 {
                moves.append(Move(square, value, disk, directions))
            }
        }
        return moves.sorted()
    }

    /// Print available move coordinates and resulting points gained.
    func printPossibleMoves(_ moves: [Move]) {
        print("  Possible moves (\(moves.count)):".yellow())
        // convert board from Disk enums to strings
        var formattedBoard = self.board.map { $0.boardChar() }
        for move in moves {
            print("  \(move)")
            formattedBoard[move.square.y * self.size + move.square.x] = "\(move.value)".yellow()
        }
        // print board with move positions
        print("   ", terminator: "")
        for x in 0 ..< self.size {
            print(" \(x)".dim(), terminator: "")
        }
        for y in 0 ..< self.size {
            print("\n  \(y)".dim(), terminator: "")
            for x in 0 ..< self.size {
                print(" \(formattedBoard[y * self.size + x])", terminator: "")
            }
        }
        print("")
    }

    /// Returns the winner color.
    func result() -> Disk {
        let sum: Int = self.score()
        switch sum {
            case ..<0:
                return Disk.black
            case 1...:
                return Disk.white
            default:
                return Disk.empty
        }
    }

    /// Print current score for both players.
    func printScore() {
        let score = self.playerScores()
        print("\n\(self)")
        print(
            "Score: \(getColor(score.black, Disk.black.color())) |",
            "\(getColor(score.white, Disk.white.color()))"
        )
    }

    /// Get board status string for game log.
    func toLogEntry() -> String {
        let logChars = self.board.map { $0.boardChar(color: false) }
        return logChars.joined()
    }

    /// Check that the given coordinates are inside the board.
    private func checkCoordinates(_ x: Int, _ y: Int) -> Bool {
        x >= 0 && x < self.size && y >= 0 && y < self.size
    }

    /// Count and return the number of black and white disks.
    private func playerScores() -> (black: Int, white: Int) {
        let counts = self.board.reduce(into: (black: 0, white: 0)) { result, disk in
            switch disk {
                case .white:
                    result.white += 1
                case .black:
                    result.black += 1
                case .empty:
                    break
            }
        }
        return counts
    }

    /// Returns the total score.
    /// Positive value means more white disks and negative means more black disks.
    private func score() -> Int {
        self.board.reduce(0) { $0 + $1.rawValue }
    }

    /// Returns the state of the board (empty, white, black) at the given square.
    private func getSquare(_ square: Square) -> Disk? {
        self.checkCoordinates(square.x, square.y) ? self
            .board[square.y * self.size + square.x] : nil
    }

    /// Sets the given square to given value.
    private func setSquare(_ square: Square, _ disk: Disk) {
        if !self.checkCoordinates(square.x, square.y) {
            print("Invalid coordinates \(square)!")
        }
        self.board[square.y * self.size + square.x] = disk
    }
}

extension Board: CustomStringConvertible {
    var description: String {
        var text = " "
        for x in 0 ..< self.size {
            text += " \(x)".dim()
        }
        for y in 0 ..< self.size {
            text += "\n\(y)".dim()
            for x in 0 ..< self.size {
                let disk = self.board[y * self.size + x]
                text += " \(disk.boardChar())"
            }
        }
        return text
    }
}
